### ISSUE TITLE
Pivot table horizontal grouping, is it supported?

### ISSUE THREAD

**Issue by hfazai** [Kotlin/dataframe#97]
Thanks for creating this great library :+1: !

I follow this [documentation](https://kotlin.github.io/dataframe/pivot.html) to create a pivot table.

Is it possible to make an horizontal grouping like in this [example](https://pivottable.js.org/examples/mps_prepop.html) (add gender to rows) ?

**Comment by koperagen** [comment]
Hi! Can you provide a screenshot of the desired table and what you get with `pivot { col1 then col2 } .groupBy { row1 and row2 }`. What i see on the link:
![image](https://user-images.githubusercontent.com/12936457/163692680-64ab66c7-ec30-4e12-856a-92022ca8913f.png)
Not sure if this is the desired table

**Comment by hfazai** [comment]
Hi, 
With `pivot { "Party" } .groupBy { "Province" and "Gender" }.sum("Age")` I want to get:
![image](https://user-images.githubusercontent.com/57005249/163718860-5a96a6bd-5dce-4226-a4c0-b1d8d8040fb4.png)

Using dataframe I get `Province `and `Gender` pivoted independently :
```
Alberta, Female, ...
.....
Alberta, Male, ...
```

**Comment by hfazai** [comment]
@koperagen Maybe we need to have the function `then` in `ColumnsSelectionDsl` like in `PivotDsl`.

**Comment by koperagen** [comment]
Thank you for the use case, it's very interesting! Maybe `df.pivot { Gender then Party }.groupBy { Province }.sum { Age }` does the thing? It yields a dataframe with this schema: 
![image](https://user-images.githubusercontent.com/12936457/163832781-a97ab6d5-499b-474c-9468-84ee9c2f7a59.png)

**Comment by hfazai** [comment]
If I have to group only two columns, `df.pivot { Gender then Party }.groupBy { Province }.sum { Age }` does the thing.
But actualy I'm using dataframe to add the Pivot Table feature to the framework [Galite](https://github.com/kopiLeft/Galite). It should allow grouping data by rows and by columns. How could I make the table below with dataframe?
![image](https://user-images.githubusercontent.com/57005249/163970156-6b181a26-0fe5-4acd-9906-92d7bc64c0f3.png)
I have gender and smoker grouped horizontally / day and time are grouped vertically.

**Comment by koperagen** [comment]
Now that you mention horizontal / vertical grouping, documentation states https://kotlin.github.io/dataframe/pivot.html 
```
df.pivot { pivotColumns }.groupBy { indexColumns }
pivotColumns — columns with values for horizontal data grouping and generation of new columns
indexColumns — columns with values for vertical data grouping
```
According to `I have gender and smoker grouped horizontally / day and time are grouped vertically.` this code should work: 
`df.pivot { smoker then gender }.groupBy { day and time }`
Does it work as you would expect? 

 Also, can you share this dataset if it's public?

**Comment by hfazai** [comment]
`df.pivot { smoker then gender }.groupBy { day and time }` Gives:
![image](https://user-images.githubusercontent.com/57005249/164001620-d88d78ed-7895-4617-8010-5cc53e7b1f62.png)

My actiual workaround is to sort Values on the row and try to span them `df.pivot { smoker then gender }.groupBy { day and time }.sortBy { day and time }` Which gives:
![image](https://user-images.githubusercontent.com/57005249/164001980-063e6a63-913d-407d-bf90-7d1b8f7abb23.png)

**Comment by koperagen** [comment]
I see the problem, yes. Can you provide the dataset so i can give it a try myself? It would be easier for me to figure it out in REPL :) Or we can pick other dataset

**Comment by nikitinas** [comment]
Hi, @hfazai. As far as I understand, you need hierarchical grouping of rows, so that `df.groupBy { a then b }` will return `DataFrame` with a single row for every distinct `a` value and every row will contain nested `DataFrame` grouped by `b`. Currently `DataFrame` supports nesting only for columns, but not for rows. It is a good feature request, thank you! We'll consider it

**Comment by nikitinas** [comment]
Currently this is the only way to perform something similar to nested rows:

<img width="1157" alt="Screen Shot 2022-04-19 at 15 53 27" src="https://user-images.githubusercontent.com/4507608/164008832-0f134be8-0f34-438a-848b-0f6174e5e96f.png">

**Comment by hfazai** [comment]
> Hi, @hfazai. As far as I understand, you need hierarchical grouping of rows, so that `df.groupBy { a then b }` will return `DataFrame` with a single row for every distinct `a` value and every row will contain nested `DataFrame` grouped by `b`. Currently `DataFrame` supports nesting only for columns, but not for rows. It is a good feature request, thank you! We'll consider it

Hi @nikitinas, yes exactly :)

**Comment by hfazai** [comment]
> Currently this is the only way to perform something similar to nested rows:
> 

I'll try it, thanks!
Actually I did a workaround with the sortBy { } and it works fine and it would be great to add this feature by adding `then` to `ColumnsSelectionDsl` so that we can call `df.groupBy { a then b }`

**Comment by nikitinas** [comment]
> My actual workaround is to sort Values on the rows and try to span them `df.pivot { smoker then gender }.groupBy { day and time }.sortBy { day and time }` which gives: ![image](https://user-images.githubusercontent.com/57005249/164001980-063e6a63-913d-407d-bf90-7d1b8f7abb23.png)

I can't match this image with your code. According to attached screenshot, you do `df.groupBy { gender and smoker  }.pivot { day then time }.sortBy { gender and smoker }`. And you want `df.groupBy { gender then smoker  }` to have all rows with the same `gender` stored consequently (by default, without sorting) and also merge equal values in `gender` vertically. Is it correct?

**Comment by nikitinas** [comment]
Let's discuss this feature. 

An obvious solution is to add `GroupedDataRow` that will store value for the first grouping key and nested `DataFrame` with other grouping keys. It will be similar to `FrameColumn` from my [previous comment](https://github.com/Kotlin/dataframe/issues/97#issuecomment-1102620092): 
<img width="244" alt="Screen Shot 2022-04-19 at 16 24 24" src="https://user-images.githubusercontent.com/4507608/164015104-dc4c43cc-4234-46f5-b021-0cfdfa32774c.png">

but will have a better presentation: nested `DataFrames` will be injected into the root table without expanders and `group` column. 

This approach raises several questions:
* How many rows should return `df.rowsCount()`? 2 or 4?
* Can `gender` and `party` columns be addressed directly as `df.gender`? If yes, will they have size 4 while `province` have size 2? This will break current restriction that all columns in `DataFrame` should have equal size.
* Should `df.sortBy { gender }` work across the whole table (break original row grouping) or sort only within row groups? 
* ...

I think that fair implementation of row grouping may overcomplicate `DataFrame` usage. I assume that in most cases grouped rows are needed only for better visual representation, so we can just add some info into `DataColumn` API with span ranges indicating subsequent values that should be rendered in a single merged cell. So internally nothing will change and such "merged" column will behave just as any other column. What do you think about it?

**Comment by nikitinas** [comment]
And we also should add `groupBy { a then b }`.

**Comment by hfazai** [comment]
> we can just add some info into `DataColumn` API with span ranges indicating subsequent values that should be rendered in a single merged cell. So internally nothing will change and such "merged" column will behave just as any other column. What do you think about it?

I agree with you concerns about changing the internal implementation. The solution of adding the additional informations about the row grouping looks good to me.
I think to add span ranges indicating subsequent values that should be rendered in a single merged cell, the `groupBy { a then b }` should sort the rows before.

**Comment by nikitinas** [comment]
I think that sorting should be performed explicitly. We can add `.asc()` and `.desc()` modifiers in column selector, similar to [`sortBy`](https://kotlin.github.io/dataframe/sortby.html):
`df.groupBy { a.asc() and b.desc() }`

**Comment by hfazai** [comment]
Sounds good :+1:

**Comment by AndreiKingsley** [comment]
Oh, the discussions we had at the start of our journey were really interesting 😄 !

But at the moment, it seems to me that KDF doesn't need support for such structure as "grouped data row".
API for basic structures (data rows/columns/group by) became stable and API for new one is hard to add now.

I haven't met any possible use-case for this. We can improve rendering for `GroupBy`, but otherwise current representation seems quite sufficient.

@zaleslaw can we close this issue?
